### PR TITLE
Add OpenMote-CC2538 platform and examples.

### PR DESCRIPTION
**Pull request by peretuset**
After a long time, the OpenMote-CC2538 platform is finally ready to be merged into the Contiki's main repository. I have tested all the platform-related code and also the examples and they all work allright.
Please review the PR and make suggestions to improve it if required. Thanks!

### CODE REVIEW

**Comment by simonduq** [comment]
Thanks the lot for bringing this back to life!!
Minor things that need to be addressed:
* fix doxygen warnings (the reason why Travis failed)
* comply with https://github.com/contiki-os/contiki/wiki/Code-Contributions#new-platforms , in particular
 * add content to the readme
 * name at least one long-term maintainer

**Comment by simonduq** [comment]
Oh I forgot one:
* add compile non-regression tests (very easy, just add a few lines to https://github.com/contiki-os/contiki/blob/master/regression-tests/18-compile-arm-ports/Makefile)

**Comment by peretuset** [comment]
After a long night banging my head against the wall with Doxygen syntax and some other minor issues, it looks like it is now passing all the CI tests! I have also added the non-regression tests and added content to the readme file. Any further suggestions are welcome!

**Comment by simonduq** [comment]
Looks great, fulfills the new platform requirements, and I know this has already been used by many. Gets my :+1:

File: platform/openmote-cc2538/Makefile.openmote-cc2538 [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+### Define the CPU directory
+CONTIKI_CPU=$(CONTIKI)/cpu/cc2538
+include $(CONTIKI_CPU)/Makefile.cc2538
+
+MODULES += core/net core/net/mac \
+           core/net/mac/contikimac \
+           core/net/llsec core/net/llsec/noncoresec
+
+PYTHON = python
+BSL_FLAGS += -e -w --bootloader-invert-lines -b 115200 -v
```
Comment (alignan) [inline platform/openmote-cc2538/Makefile.openmote-cc2538]: It is not mandatory, but the new `cc2538-bsl` allows higher baudrates than 115200, for other platforms the `-b` argument was dropped

File: platform/openmote-cc2538/contiki-conf.h [hunk 2]
```diff
@@ -0,0 +1,573 @@
...
+#endif
+/** @} */
+
+/*---------------------------------------------------------------------------*/
+/**
+ * \name CFS configuration
+ *
+ * @{
+ */
+#define COFFEE_CONF_SIZE            (4 * COFFEE_SECTOR_SIZE)
```
Comment (alignan) [inline platform/openmote-cc2538/contiki-conf.h]: it was suggested by @bthebaudeau to change this to:

````
#ifndef COFFEE_CONF_SIZE
#define COFFEE_CONF_SIZE            (4 * COFFEE_SECTOR_SIZE)
#endif
````

File: platform/openmote-cc2538/dev/adxl346.c [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+  i2c_single_receive(ADXL346_ADDRESS, &acceleration[0]);
+  i2c_single_send(ADXL346_ADDRESS, ADXL346_DATAZ1_ADDR);
+  i2c_single_receive(ADXL346_ADDRESS, &acceleration[1]);
+
+  z = (acceleration[0] << 8) | acceleration[1];
+
+  return z;
+}
+/*---------------------------------------------------------------------------*/
+/** @} */
```
Comment (alignan) [inline platform/openmote-cc2538/dev/adxl346.c]: No newline at the end of the file

File: platform/openmote-cc2538/dev/max44009.c [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+  exponent = (exponent == 0x0F ? exponent & 0x0E : exponent);
+  
+  mantissa = (lux >> 0) & 0xFF;
+
+  result *= 2 ^ exponent * mantissa;
+
+  return result;
+}
+/*---------------------------------------------------------------------------*/
+/** @} */
```
Comment (alignan) [inline platform/openmote-cc2538/dev/max44009.c]: No newline at the end of the file

File: platform/openmote-cc2538/dev/max44009.h [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+ * \brief Convert the light from the MAX44009 sensor
+ */
+float max44009_convert_light(uint16_t light);
+/*---------------------------------------------------------------------------*/
+#endif
+/*---------------------------------------------------------------------------*/
+/**
+ * @}
+ * @}
+ */
```
Comment (alignan) [inline platform/openmote-cc2538/dev/max44009.h]: Missing newline

File: platform/openmote-cc2538/dev/sht21.c [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+{
+  float result;
+
+  result = -6.0;
+  result += 125.0 * (float) humidity / 65536.0;
+
+  return result;
+}
+/*---------------------------------------------------------------------------*/
+/** @} */
```
Comment (alignan) [inline platform/openmote-cc2538/dev/sht21.c]: Newline missing

File: platform/openmote-cc2538/dev/sht21.h [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+ * \brief Convert the relative humidity from the SHT21 sensor
+ */
+float sht21_convert_humidity(uint16_t humidity);
+/*---------------------------------------------------------------------------*/
+#endif
+/*---------------------------------------------------------------------------*/
+/**
+ * @}
+ * @}
+ */
```
Comment (alignan) [inline platform/openmote-cc2538/dev/sht21.h]: Newline missing

**Comment by alignan** [comment]
It would be better and try to comply with Contiki's sensor API (`configure`, `value`), even if it needs a bit of rework (in my opinion), then have extended functions for features that do not fit the API.

**Comment by g-oikonomou** [comment]
Good to see activity here again.

* There are a number of code style inconsistencies in various files, please run this through uncrustify.
* The doxygen documentation is terribly broken. There are 3 `\defgroup openmote-cc2538`. There should be only one and it should be added to group `platform`. I can see many `\addtogroup openmote`, but I cannot see the respective `defgroup` anywhere. The outcome is that the platforms documentation modules are scattered all across the html docs. This should be tidied up.
* As Antonio says, unless there is a specific reason to force speed 115200 on the BSL, allowing the default of 1000000 is recommended for faster programming.

**Comment by g-oikonomou** [comment]
Furthermore, the convention for include guards is `THIS_H_`, so some of those need changed in this pull (don't forget the closing comment). We renamed across the board those that started with `__` a while ago and we should try to not introduce new ones.

File: platform/openmote-cc2538/dev/antenna.c [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+  /* Select external antenna by default. */
+  antenna_external();
+}
+/*---------------------------------------------------------------------------*/
+void
+antenna_external(void)
+{
+  gpio_reset(BSP_RADIO_BASE, BSP_RADIO_INT);
+  gpio_set(BSP_RADIO_BASE, BSP_RADIO_EXT);
```
Comment (drandreas) [inline platform/openmote-cc2538/dev/antenna.c]: In my humble opinion ```GPIO_WRITE_PIN``` should be used here.

File: platform/openmote-cc2538/Makefile.openmote-cc2538 (hunk 1 above)
Comment (peretuset) [inline platform/openmote-cc2538/Makefile.openmote-cc2538]: Indeed, thanks!

File: platform/openmote-cc2538/contiki-conf.h (hunk 2 above)
Comment (peretuset) [inline platform/openmote-cc2538/contiki-conf.h]: Noted, thanks!

File: platform/openmote-cc2538/dev/adxl346.c (hunk 3 above)
Comment (peretuset) [inline platform/openmote-cc2538/dev/adxl346.c]: Done!

File: platform/openmote-cc2538/dev/max44009.c (hunk 4 above)
Comment (peretuset) [inline platform/openmote-cc2538/dev/max44009.c]: Done!

File: platform/openmote-cc2538/dev/max44009.h (hunk 5 above)
Comment (peretuset) [inline platform/openmote-cc2538/dev/max44009.h]: Done!

File: platform/openmote-cc2538/dev/sht21.c (hunk 6 above)
Comment (peretuset) [inline platform/openmote-cc2538/dev/sht21.c]: Done!

File: platform/openmote-cc2538/dev/sht21.h (hunk 7 above)
Comment (peretuset) [inline platform/openmote-cc2538/dev/sht21.h]: Done!

File: platform/openmote-cc2538/dev/antenna.c (hunk 8 above)
Comment (peretuset) [inline platform/openmote-cc2538/dev/antenna.c]: Done, thanks!

**Comment by peretuset** [comment]
Just got through all of your comments and the checks are still passing. Can you double check that the issues are resolved now? Thanks!

File: platform/openmote-cc2538/dev/max44009.h [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+ *
+ * This file is part of the Contiki operating system.
+ *
+ */
+/*---------------------------------------------------------------------------*/
+/**
+ * \addtogroup openmote-sensors
+ * @{
+ *
+ * \defgroup openmote-max4009-sensor MAX4009 light sensor
```
Comment (g-oikonomou) [inline platform/openmote-cc2538/dev/max44009.h]: Due to a typo (should be 44009 - one '4' missing), this ends up in the wrong place in the doc module hierarchy

**Comment by g-oikonomou** [comment]
Pere, thanks. Other than the typo that I flagged inline, all my comments have been addressed and I'd be more than happy to merge this one (at long last!) :+1:

File: platform/openmote-cc2538/dev/max44009.h (hunk 9 above)
Comment (peretuset) [inline platform/openmote-cc2538/dev/max44009.h]: Fixed the typo, should be ready to merge once the test are passed.
Thanks for taking care of the review!

**Comment by g-oikonomou** [comment]
:+1:

**Comment by alignan** [comment]
I just have a comment about making the sensors compatible with current Contiki sensors implementation, but no strong objection about it (it can be done in a new PR, although this normally doesn't happen)... @g-oikonomou mind-tricked me to port all of our sensors a while ago, so perhaps he can comment :smile_cat:

**Comment by g-oikonomou** [comment]
Even though the sensors API is a little PITA, I always choose to use it and encourage others to do so. When I read the comments I somehow thought this has been done already. Would be nice, but I'm not hugely bothered - support for this platform is long overdue

**Comment by kkrentz** [comment]
:+1: Works, except for some issue with the 6LoWPAN adaption layer. A small change did the trick: https://github.com/kkrentz/contiki/commit/eec004d796f47d9dfc28a39ee95ae20b9aeab1bb

**Comment by alignan** [comment]
@kkrentz in any case, it should be:
````
/* Define our IPv6 prefixes/contexts here */
#define SICSLOWPAN_CONF_MAX_ADDR_CONTEXTS    1
#ifndef SICSLOWPAN_CONF_ADDR_CONTEXT_0
#define SICSLOWPAN_CONF_ADDR_CONTEXT_0 { \
  addr_contexts[0].prefix[0] = UIP_DS6_DEFAULT_PREFIX_0; \
  addr_contexts[0].prefix[1] = UIP_DS6_DEFAULT_PREFIX_1; \
}
#endif
````
Am I wrong? or is there any other related problem for other platforms as well?

**Comment by kkrentz** [comment]
This also works - so apparently the problem I experienced was a wrong mapping between context identifier and network prefix

**Comment by g-oikonomou** [comment]
@peretuset This pull needs updated to reflect recent changes to 6LoWPAN contexts, as per the discussion above. The preferred approach is the one Antonio mentions, also see #1297

Also, I need some comment on whether you have any plans of updating sensor drivers to use the standard API.

**Comment by peretuset** [comment]
Done!

**Comment by g-oikonomou** [comment]
I'm merging this. It is still advisable to update sensor drivers to use the Contiki sensors API. But then again, this hardware and respective support have been around for too long already!

**Comment by peretuset** [comment]
Super, thanks!
By the way, I have finished moving the sensors drivers to use the Contiki sensors API, so you can expect a new PR soon enough... :)

**Pull request merged**